### ISSUE TITLE
Any chance the plugin could work with Typescript?

### ISSUE THREAD

**Issue by domehead100** [gajus/babel-plugin-react-css-modules#130]
I've been using this plugin for about 9 months to great effect.  Now we're trying to start introducing Typescript, so using the ts-loader instead of the babel-loader for js/ts files.  I am getting my css generated with the scoped names on the classes, but my React components are not getting transformed, so I'm ending up with "stylename" tags remaining in the output js and then React (<16) is complaining about unknown stylename attributes on the components.

I tried chaining babel-loader from ts-loader, but that had no effect.

My .babelrc:
```javascript
{
  "presets": [["env", { modules: false}]],
  "plugins": [
    ["react-css-modules", { "generateScopedName": "[name]__[local]" }],
  ]
}
```
My loader for js/ts:
```javascript
 // ts/js
{
    test: /\.(t|j)sx?$/,
    include: PATHS.src,
    use: [
        // { // babel settings in in .babelrc file
        //     loader: "babel-loader",
        // },
        {
            loader: "ts-loader",
        },
    ],
},
```
And my loader for css:
```javascript
// non-global css (e.g., css-modules)
// Note that we're not using style-loader, which appends style tag(s) to the html body,
// which we really don't ever want
{
    test: /\.css$/,
    exclude: /global\.css$/,
    loader: ExtractTextPlugin.extract({
        use: [
            // load component css via css modules with local scoping
            {
                loader: "css-loader",
                query: {
                    modules: true,
                    importLoaders: 1,
                    localIdentName: "[name]__[local]",
                    minimize: false,
                    sourceMap: true,
                },
            },
            "postcss-loader", // config is in postcss.config.js
        ],
    }),
},
```

**Comment by domehead100** [comment]
Some progress, switching the value for the "jsx" property of the typescript compiler options from "react" to "preserve" now seems to have the plugin being called if I chain ts-loader and then babel-loader.  Now having an error in a file where I have an anonymous import and a named import.  

import "./globals.css"
import app from "./app.css"
or
import * as app from "./app.css"
or
let app = require<any>("./app.css")

The plugin is telling me that the "CSS module import does not exist".  Putting a breakpoint shows that globals is in the styleModuleImportMap, but not the app.  

Trying to figure this out...

**Comment by domehead100** [comment]
Closing this since everything works after some poking around.

To summarize for posterity... you need to chain the typescript loader (I'm using ts-loader) and the babel loader for the plugin to be invoked.

You need to use es6 import format for anonymous and named imports, e.g.:
```javascript
import './component.css'
import app from './app.css'
```
If you have a named import, you need to reference the variable somewhere, otherwise the typescript compiler will strip it before babel gets a look at it, e.g.:
```javascript
let dummy = app
```
Then stylename attributes work:
```jsx
<div syleName="_" /> // from component.css
<div styleName="app.somestyle" /> // from app.css
```

loader for js/jsx/ts/tsx in webpack.config.js:
```javascript
// ts/js
{
    test: /\.(t|j)sx?$/,
    include: PATHS.src,
    use: [
        { // babel settings in in .babelrc file
            loader: "babel-loader",
        },
        {
            loader: "ts-loader",
        },
    ],
},
```
loader for css that is to be processed by babel-plugin-react-css-modules:
```javascript
// non-global css (e.g., css-modules)
// Note that we're not using style-loader, which appends style tag(s) to the html body,
// which we really don't ever want
{
    test: /\.css$/,
    exclude: /global\.css$/,
    loader: ExtractTextPlugin.extract({
        use: [
            // load component css via css modules with local scoping
            {
                loader: "css-loader",
                query: {
                    modules: true,
                    importLoaders: 1,
                    localIdentName: "[name]__[local]",
                    minimize: false,
                    sourceMap: true,
                },
            },
            "postcss-loader", // config is in postcss.config.js
        ],
    }),
},
```
.babelrc:
```javascript
// the react-css-modules plugin should be first in the list of plugins
{
  "presets": [["env", { "modules": false}], "react"],
  "plugins": [
    ["react-css-modules", { "generateScopedName": "[name]__[local]" }]
  ]
}
```
a type file that is referenced by typescript, mine is called typings.d.ts:
```javascript
declare module "*.css" {
    interface IClassNames {
        [className: string]: string;
    }
    const classNames: IClassNames;
    export = classNames;
}
```

**Comment by atdrago** [comment]
Thank you @domehead100! This helped me a ton!

**Comment by zvzuola** [comment]
@domehead100 I tried you resolution，but it does not work with me，the error is "Cannot use styleName attribute without importing at least one stylesheet.
    at Array.map (<anonymous>)"
where is wrong?
here is my config:
`{
            test: /\.tsx?$/,
            include: paths.appSrc,
            use: [{
              loader: require.resolve('babel-loader'),
              options: {
                plugins: [
                  ["react-css-modules", {
                    "context": "./",
                    "generateScopedName": "[local]-[hash:base64:15]"
                  }]
                ],
                cacheDirectory: true,
              },
            }, {
              loader: 'awesome-typescript-loader',
            }]
          },`
and my .tsx file:
`import * as React from 'react'
const style = require('./Index.css')
// import style from './Index.css'
export default class Invitation extends React.Component {
  render() {
    return (
      <div styleName='aa'>
        test
      </div>
    )
  }
}`
tsconfig.json:
`{
    "compilerOptions": {
        "outDir": "./build/",
        "sourceMap": true,
        "noImplicitAny": true,
        "module": "commonjs",
        "target": "es5",
        "jsx": "preserve",
        "allowJs": true
    },
    "include": [
        "./src/**/*"
    ]
}`

**Comment by henrik-sporta** [comment]
Agree with previous comments, truly helpful!

Although, I still struggled for a while, until I changed an option in ts-loader

```
{
    loader: "ts-loader",
    options: {
        transpileOnly: true
    }
},
```
which took me a couple of steps further in the mess.

**Comment by hipstersmoothie** [comment]
@domehead100 how did you get the type checking working? styleName throws and error for me

**Comment by izikandrw** [comment]
Has anyone found a way to get this to work without having to assign the imported css to a dummy variable--I feel like there should be a way to tell babel to leave seemingly unused imports alone.

**Comment by henrik-sporta** [comment]
> Has anyone found a way to get this to work without having to assign the imported css to a dummy variable--I feel like there should be a way to tell babel to leave seemingly unused imports alone.

If I remember correctly, babel isn't the problem, it's`ts-loader`. 

After I've used this for a while now and also tried out other options in other projects, my recommendation is to simply go with `css-loader` and skip this completely. Yeah, you lose the ability to use the `styleName` attribute but that doesn't really give you that much. Being limited to use style modules with `className` shouldn't be a bottleneck for anyone (at least from my experience). 

Besides, by using this you're adding another complexity layer to your bundler. I ran it some awkward challenges when I wanted to manipulate the resulting hash (`generateScopedName`/`localIdentName`) since it was handled in two places (`react-css-modules` and `css-loader`) instead of just one.
